### PR TITLE
New Login system

### PR DESCRIPTION
**Pull request by NickHeiner**
Connect #2272.

Run `rake db:seed` to get the test user. Once you do this, you'll see `Reader at station 405`, which is what you need to use for testing.

2272 login

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -9,18 +9,24 @@ def new
...
     end
 
-    return redirect_to(ENV["SSO_URL"]) unless current_user
-  end
-
-  def create
-    unless current_user.authenticate(authentication_params)
-      flash[:error] = "The username and password you entered don't match, please try again."
-      return redirect_to login_path
+    if current_user.ro_is_ambiguous_from_station_office?
```
Comment (aroltsch) [inline app/controllers/sessions_controller.rb]: Could you please add a comment here why we are doing it. It is a bit unclear.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -9,18 +9,24 @@ def new
...
+      @regional_office_options = current_user.station_offices.map do |regional_office_code|
+        {
+          "regionalOfficeCode" => regional_office_code,  
+          "regionalOffice" => VACOLS::RegionalOffice::CITIES[regional_office_code]
+        }
+      end
+      @redirect_to = session["return_to"] || root_path
+    else
+      return redirect_to(ENV["SSO_URL"]) unless current_user
```
Comment (aroltsch) [inline app/controllers/sessions_controller.rb]: if `current_user` can be nil, your code on line 12 will break.

File: app/services/authentication_service.rb [hunk 3]
```diff
@@ -4,6 +4,7 @@ def self.default_user_session
   end
 
   def self.authenticate_vacols(regional_office, password)
+    puts "try to auth VACOLS ================="
```
Comment (aroltsch) [inline app/services/authentication_service.rb]: I think you might have forgotten to remove this :)

File: app/models/user.rb [hunk 4]
```diff
@@ -65,8 +69,7 @@ def authenticated?
 
   # This method is used for VACOLS authentication
   def authenticate(regional_office:, password:)
-    return false unless User.authenticate_vacols(regional_office, password)
-
+    # return false unless User.authenticate_vacols(regional_office, password)
```
Comment (aroltsch) [inline app/models/user.rb]: Did you mean to comment this out?

File: app/models/user.rb [hunk 5]
```diff
@@ -25,9 +25,13 @@ def roles
     end
   end
 
-  # If RO is unambiguous from station_office, use that RO. Otherwise, use user defined RO
+  # If RO is ambiguous from station_office, use the user-defined RO. Otherwise, use the unambigous RO.
   def regional_office
-    station_offices.is_a?(String) ? station_offices : @regional_office
+    ro_is_ambiguous_from_station_office? ? @regional_office : station_office
```
Comment (aroltsch) [inline app/models/user.rb]: Where is `station_office` coming from? I can only see `station_offices` or `station_id` on the User model.

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -9,18 +9,31 @@ def new
       return render "errors/500", layout: "application", status: 503
     end
 
-    return redirect_to(ENV["SSO_URL"]) unless current_user
-  end
+    unless current_user
```
Comment (aroltsch) [inline app/controllers/sessions_controller.rb]: You could make this a one liner:
```
return redirect_to(ENV["SSO_URL"]) unless current_user
```

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -9,18 +9,31 @@ def new
...
+    end
 
-  def create
-    unless current_user.authenticate(authentication_params)
-      flash[:error] = "The username and password you entered don't match, please try again."
-      return redirect_to login_path
+    # In order to use Caseflow, we need to know what regional office (RO) the user is from.
+    # CSS will give us the station office ID. Some station office IDs correspond to multiple
+    # RO IDs. In this case, we present a list of ROs to the user and ask which one they are.
+    if current_user.ro_is_ambiguous_from_station_office?
```
Comment (aroltsch) [inline app/controllers/sessions_controller.rb]: I am personally not a big fan of if/else statements :)
You could do it like this:
```
unless current_user.ro_is_ambiguous_from_station_office?
   redirect_to(session["return_to"] || root_path)
   return
end
@regional_office_options = current_user.station_offices.map do |regional_office_code|
...
end
@redirect_to = session["return_to"] || root_path
```

**Comment by aroltsch** [comment]
Hey Nick,
I don't see any tests as part of this PR. Would existing tests do the job for these new code changes?

**Comment by NickHeiner** [comment]
I am adding new tests as we speak. Thanks for calling that out!

File: client/app/components/DropDown.jsx [hunk 8]
```diff
@@ -40,7 +40,7 @@ DropDown.propTypes = {
   label: PropTypes.string,
   name: PropTypes.string.isRequired,
   onChange: PropTypes.func,
-  options: PropTypes.array,
+  options: PropTypes.array.isRequired,
```
Comment (tejans24) [inline client/app/components/DropDown.jsx]: This component doesn't seem like it's being used anywhere in the codebase. Maybe we should just deprecate/remove it? This might be out of scope for this PR.

Comment (NickHeiner) [inline client/app/components/DropDown.jsx]: Good point – let's do that in another PR.

File: client/app/login/index.jsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    this.setState({ isLoggingIn: true });
+    ApiUtil.patch('/sessions/update', {
+      data: {
+        regional_office: this.state.regionalOfficeCode
+      }
+    }).then(
+      () => window.location = this.props.redirectTo,
+      (err) => {
+        this.setState({ isLoggingIn: false });
+        // eslint-disable-next-line no-console
```
Comment (tejans24) [inline client/app/login/index.jsx]: Does this need an error message when login call fails? It looks like it was mentioned in the comments of #2272

Comment (NickHeiner) [inline client/app/login/index.jsx]: Indeed. I put this in before I had that spec. I'd like to do that as a separate PR, since we are getting close to the deadline for this.

File: client/app/login/index.jsx [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        name="Log in"
+        loading={this.state.isLoggingIn}
+        loadingText="Logging in" />
+    </div>;
+  }
+}
+
+Login.propTypes = {
+  regionalOfficeOptions: PropTypes.array.isRequired
+};
```
Comment (tejans24) [inline client/app/login/index.jsx]: redirectTo can be added to propTypes

File: client/app/login/index.jsx (hunk 9 above)
Comment (NickHeiner) [inline client/app/login/index.jsx]: Actually, I put it into this PR.

File: spec/feature/login_spec.rb [hunk 11]
```diff
@@ -14,7 +13,7 @@
     Rails.application.config.sso_service_disabled = false
   end
 
-  scenario "User who's station ID has one RO doesn't require login" do
+  scenario "User whose station ID has one RO doesn't require login" do
```
Comment (joofsh) [inline spec/feature/login_spec.rb]: 🙌

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -10,17 +10,29 @@ def new
...
+      end
+      @redirect_to = session["return_to"] || root_path
+      return
     end
 
+    redirect_to(session["return_to"] || root_path)
+  end
+
+  def update
+    current_user.regional_office = params["regional_office"].upcase
```
Comment (joofsh) [inline app/controllers/sessions_controller.rb]: We should get rid of the `authentication_params` on old line 34. Maybe replace with `regional_office_params`.

Also, can the param be nil? Will the `nil.upcase` potentially explode?

File: config/routes.rb [hunk 13]
```diff
@@ -89,6 +89,7 @@
 
   get "health-check", to: "health_checks#show"
   get "login" => "sessions#new"
+  get "sessions/update" => "sessions#update"
```
Comment (joofsh) [inline config/routes.rb]: what do we need this get route for?

File: app/models/user.rb [hunk 14]
```diff
@@ -107,8 +106,6 @@ def to_hash
     serializable_hash
   end
 
-  private
```
Comment (aroltsch) [inline app/models/user.rb]: Did you mean to remove it completely from this class?

File: app/models/user.rb [hunk 15]
```diff
@@ -25,9 +25,15 @@ def roles
     end
   end
 
-  # If RO is unambiguous from station_office, use that RO. Otherwise, use user defined RO
+  # If RO is ambiguous from station_office, use the user-defined RO. Otherwise, use the unambigous RO.
   def regional_office
-    station_offices.is_a?(String) ? station_offices : @regional_office
+    upcase = ->(str) { str ? str.upcase : str }
+
+    ro_is_ambiguous_from_station_office? ? upcase.call(@regional_office) : station_offices
```
Comment (aroltsch) [inline app/models/user.rb]: Is there a test for this line?

**Comment by aroltsch** [comment]
Nick, one more thing I'd like to mention. How is eFolder Express login gonna be affected?

File: app/models/user.rb (hunk 14 above)
Comment (NickHeiner) [inline app/models/user.rb]: Yes. `station_offices` was made public, and there was nothing else that was `private`.

File: app/models/user.rb (hunk 15 above)
Comment (NickHeiner) [inline app/models/user.rb]: Yes.

```rb
  context "#authenticate" do
    subject { user.regional_office = "rO21" }

    context "when user enters lowercase RO" do
      it "sets regional_office in the session" do
        is_expected.to be_truthy
        expect(user.regional_office).to eq("RO21")
      end
    end
  end
```
